### PR TITLE
Querify `FnAbi::of_{fn_ptr,instance}` as `fn_abi_of_{fn_ptr,instance}`.

### PR DESCRIPTION
**Pull request by eddyb**
*Note: opening this PR as draft because it's based on #88499*

This more or less replicates the `LayoutOf::layout_of` setup from #88499, to replace `FnAbi::of_{fn_ptr,instance}` with `FnAbiOf::fn_abi_of_{fn_ptr,instance}`, and also route them through queries (which `layout_of` has used for a while).

The two changes at the use sites (other than the names) are:
* return type is now wrapped in `&'tcx`
  * the value *is* interned, which may affect performance
* the `extra_args` list is now an interned `&'tcx ty::List<Ty<'tcx>>`
  * should be cheap (it's empty for anything other than C variadics)
 
Theoretically, a `FnAbiOfHelpers` implementer could choose to keep the `Result<...>` instead of eagerly erroring, but the only existing users of these APIs are codegen backends, so they don't (want to) take advantage of this.
At least miri could make use of this, since it prefers propagating errors (it "just" doesn't use `FnAbi` yet - cc @RalfJung).

The way this is done is probably less efficient than what is possible, because the queries handle the correctness-oriented API (i.e. the split into `fn` pointers vs instances), whereas a lower-level query could end up with more reuse between different instances with identical signatures.

r? @nagisa cc @oli-obk @bjorn3

### CODE REVIEW

**Comment by eddyb** [comment]
Thanks to https://github.com/rust-lang/rust/pull/88499#issuecomment-908688904 (i.e. #88499 being perf-neutral), we should be able to get useful data despite this being based on #88499:

@bors try @rust-timer queue

**Comment by RalfJung** [comment]
> At least miri could make use of this, since it prefers propagating errors (it "just" doesn't use FnAbi yet - cc @RalfJung).

Yeah, it's on my list... 
For LayoutOf, I think we have to properly propagate errors or some CTFE or pattern matching code will end up ICEing. Not sure if the same concern applies here.

**Comment by bjorn3** [comment]
Unfortunately this is a clear perf regression.

**Comment by eddyb** [comment]
Trying without interning, maybe it's a waste of time (tho now there would be cloning, so that's not reassuring):
@bors try @rust-timer queue

**Comment by eddyb** [comment]
Okay, so it's not the interning, at least not to a significant extent.
Next up, `eval_always` (which I'm not sure what it will do, but if it turns off caching, that'd be useful to know the effects of):

@bors try @rust-timer queue

**Comment by bjorn3** [comment]
`eval_always` doesn't turn off caching within a compilation session. It only disables caching across incremental compilation sessions. There is also `anon` which causes no new dep node to be registered, but I believe that still caches.

**Comment by bjorn3** [comment]
This effectively disabled most of incremental compilation it seems by causing all codegen units to always be dirty.

**Comment by eddyb** [comment]
Looking again at the first results (https://github.com/rust-lang/rust/pull/88575#issuecomment-911207024), something is weird: there's a lot of `-check` regressions. I wonder if they're affected specifically by the number of queries, or their key types (size or something else).

**Comment by eddyb** [comment]
Opened #88705 to test a hunch on its own (that the queries existing, even without being used, or even having any providers, can slow things down).

**Comment by eddyb** [comment]
#88705 was either a dead end, or I'm missing some aspect.

Going to try now without the last commit (or rather, with a revert of it), which actually querifies:
@bors try @rust-timer queue

**Comment by eddyb** [comment]
Further combining a lack of queries with a lack of interning, to try and eliminate the full perf difference this PR could produce:
@bors try @rust-timer queue

**Comment by eddyb** [comment]
Now that I got more `-check` regressions with an even more reduced version of the PR, I've opened #88740 for the `intern_ref` -> `intern` change.

If that turns out to be the culprit, I may be able to get away with only arena-allocating `FnAbi`s, instead of interning them.

**Comment by eddyb** [comment]
Maybe the problem with the last run is that it didn't fully remove the `FnAbi` support from the `tcx` arena and interners (it was just unused), so trying again with it removed:

@bors try @rust-timer queue

**Comment by eddyb** [comment]
Alright, that's at least part of the regression tracked down - I need to look at the arena infrastructure to see why unused arenas could cause perf changes - hopefully this isn't something about the size of the `GlobalCtxt` `struct` (backing `TyCtxt`), though that could have workarounds too.

**Comment by eddyb** [comment]
Looks like only interners change the size of `GlobalCtxt`, not arenas (and I can't find any interactions *between* arenas - though if I mark the `FnAbi` as `few` in the macro, I could also remove its space overhead there, since it would use the more dynamic "drop arena").

I've reinstated the queries, but only with arena-allocating, no interning - hopefully this won't impact `-check` benchmarks anymore.

@bors try @rust-timer queue

**Comment by eddyb** [comment]
While this isn't perfect yet, this is much closer to what I wanted to see: a lot of improvements (albeit small), and some regressions where the reuse benefits don't apply, but e.g. incremental tracking gets a bit more overhead.

In fact, I just looked at the query view for the dozen worst regressions, and they *all* have a large increase in `incr_comp_encode_dep_graph`. So maybe it's time to try that `anon` option @bjorn3 mentioned (in https://github.com/rust-lang/rust/pull/88575#issuecomment-911638052).

**Comment by eddyb** [comment]
Now with `anon` queries:

@bors try @rust-timer queue

**Comment by eddyb** [comment]
`anon` doesn't have the impact I was hoping for, that looks more or less identical.

Overall I'm pretty baffled that reserializing the *unmodified* (for `incr-unchanged` at least) dep-graph is the apparent source of the regression, and not the deserializing side.

Are small (or no) changes to the dep-graph inefficiently handled for now, and every query we add comes at a price? (cc @mw @wesleywiser)

**Comment by bjorn3** [comment]
> Are small (or no) changes to the dep-graph inefficiently handled for now, and every query we add comes at a price?

The dep graph is always reserialized from scratch. Small changes will require this in any case as serialization turns the dep node list from a list that may have holes (in case dep nodes were no longer used in the current incr comp session) to a dense list without holes, thus requiring remapping of dep node indexes.

**Comment by eddyb** [comment]
That sounds like a compacting GC, I believe those have thresholds after which they engage, so that small holes don't cause these kinds of latency issues.

**Comment by nagisa** [comment]
@eddyb you want to move this out of draft state?

**Comment by eddyb** [comment]
> @eddyb you want to move this out of draft state?

Done, I'm just worried we can't land it because of the incr-unchanged perf regressions.

File: compiler/rustc_middle/src/ty/layout.rs [hunk 1]
```diff
@@ -2796,9 +2808,39 @@ pub fn conv_from_spec_abi(tcx: TyCtxt<'_>, abi: SpecAbi) -> Conv {
...
+}
+
+impl<'tcx> fmt::Display for FnAbiError<'tcx> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Layout(err) => err.fmt(f),
+            Self::AdjustForForeignAbi(err) => err.fmt(f),
+        }
+    }
+}
```
Comment (nagisa) [inline compiler/rustc_middle/src/ty/layout.rs:2832]: Typically I would flag this with a request to implement `std::error::Error` and its `source` method, as well as to add some additional information at each layer. In my experience with `thiserror`, its `#[transparent]` errors (which this seems a lot like it) are pretty likely to produce hard to understand errors.

Seems fine in this particular instance though.

File: compiler/rustc_middle/src/ty/context.rs [hunk 2]
```diff
@@ -129,9 +129,9 @@ pub struct CtxtInterners<'tcx> {
     projs: InternedSet<'tcx, List<ProjectionKind>>,
     place_elems: InternedSet<'tcx, List<PlaceElem<'tcx>>>,
     const_: InternedSet<'tcx, Const<'tcx>>,
-    /// Const allocations.
-    allocation: InternedSet<'tcx, Allocation>,
+    const_allocation: InternedSet<'tcx, Allocation>,
```
Comment (nagisa) [inline compiler/rustc_middle/src/ty/context.rs:132]: Nit: This asks for `const_allocations` (it looks like there may be more than 1 of them?)

**Review by nagisa — COMMENTED** [review]
note to self: 4 commits remain to look at.

File: compiler/rustc_middle/src/ty/context.rs (hunk 2 above)
Comment (eddyb) [reply compiler/rustc_middle/src/ty/context.rs:132]: The interner fields are the names of the interned type, so plurals only show up for interned lists.

File: compiler/rustc_middle/src/query/mod.rs [hunk 3]
```diff
@@ -1136,6 +1136,7 @@ rustc_queries! {
         key: ty::ParamEnvAnd<'tcx, (ty::PolyFnSig<'tcx>, &'tcx ty::List<Ty<'tcx>>)>
     ) -> Result<&'tcx abi::call::FnAbi<'tcx, Ty<'tcx>>, ty::layout::FnAbiError<'tcx>> {
         desc { "computing call ABI of `{}` function pointers", key.value.0 }
+        anon
```
Comment (nagisa) [inline compiler/rustc_middle/src/query/mod.rs:1139]: Is this entirely for performance purposes? It sounds to me like it should be undesirable to enable this.

**Comment by nagisa** [comment]
I don't see any problems from the implementation standpoint here. From the discussion above it also seems like the regressions are largely caused by the implementation of the query system itself?

If so, I could see ourselves arguing that the more queryfied the compiler is, the easier it will be for us to see what the parts of the query system are a lucurative optimization target and spend time on those accordingly.

Basically the local minima vs global minima thing.

File: compiler/rustc_middle/src/query/mod.rs (hunk 3 above)
Comment (eddyb) [reply compiler/rustc_middle/src/query/mod.rs:1139]: Hmm, did I keep a HACK comment around? This was for testing perf and can be removed.

Comment (nagisa) [reply compiler/rustc_middle/src/query/mod.rs:1139]: There's a HACK in the commit message.

**Comment by oli-obk** [comment]
Yea, imo we can merge this if we have an issue for the incremental-unchanged regression that somewhat collects the information from this discussion

**Comment by wesleywiser** [comment]
+1, seems ok to merge to me.

**Comment by pnkfelix** [comment]
Visited for weekly performance triage. Results do indeed seem moderately mixed.

**Comment by eddyb** [comment]
@pnkfelix It's a bit more subtle than that (only `incr-{unchanged,patched}` regressed), and it got discussed before landing (in fact I personally wasn't sure this could land any time soon), see https://github.com/rust-lang/rust/pull/88575#issuecomment-921912025.
